noir-projects/noir-protocol-circuits/crates/rollup-lib/src/base/base_rollup_inputs.nr

```diff
@@ -294,61 +293,59 @@ impl BaseRollupInputs {
     // from the balance of the fee_payer, using the fee_payer_fee_juice_balance_read_hint to read the current balance.
     // Returns the data update request that subtracts the tx_fee from the fee_payer's balance, and the index where it
     // should be inserted in the public data update requests array.
-    fn build_or_patch_payment_update_request(
-        self,
-        tx_fee: Field,
-    ) -> (PublicDataUpdateRequest, u32) {
+    fn build_or_patch_payment_update_request(self, tx_fee: Field) -> (PublicDataWrite, u32) {
         let fee_payer = self.kernel_data.public_inputs.fee_payer;
 
         // TODO(@spalladino) Eventually remove the is_zero condition as we should always charge fees to every tx
         if !fee_payer.is_zero() {
             let read_hint = self.fee_payer_fee_juice_balance_read_hint;
             let leaf_slot = compute_fee_payer_fee_juice_balance_leaf_slot(fee_payer);
+            let existing_update_index = unsafe {
+                find_index_hint(
+                    self.kernel_data.public_inputs.end.public_data_writes,
+                    |w: PublicDataWrite| w.leaf_slot == leaf_slot,
+                )
+            };
 
-            if read_hint.leaf_slot == 0 {
+            if existing_update_index != MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX {
```

Comment:
Use the `existing_update_index` to decide whether to update the pending set or the tree.

This makes the code in the orchestrator simpler. It doesn't have to check whether the write exists in the pending set to generate the hint. It can always generate the hint. And let the circuit decide whether to use the hint or not.